packages/ckbtc/src/utils/minter.utils.ts

```diff
@@ -0,0 +1,37 @@
+import { Cbor } from "@dfinity/agent";
+
+// The minter converted a single UTXO to ckBTC.
+export const MINT_MEMO_TYPE_UTXO_TO_CKBTC = 0;
+// The minter minted accumulated KYT fees to the KYT provider.
+export const MINT_MEMO_TYPE_ACCUMULATED_KYT_FEES = 1;
+// The minter failed to check retrieve btc destination address or the destination address is tainted.
+export const MINT_MEMO_TYPE_KYT_FAIL = 2;
+
+// The memo will decode to either:
+// * Convert: [0, [ tx_id, vout, kyt_fee]]
+// * Kyt: [1]
+// * KytFail: [2, [ kyt_fee, kyt_status, block_index]]
+// Source: https://github.com/dfinity/ic/blob/c22a5aebd4f26ae2e4016de55e3f7aa00d086479/rs/bitcoin/ckbtc/minter/src/memo.rs#L25
+export type MintMemo =
+  | [typeof MINT_MEMO_TYPE_UTXO_TO_CKBTC, [Uint8Array?, number?, number?]]
+  | [typeof MINT_MEMO_TYPE_ACCUMULATED_KYT_FEES]
+  | [typeof MINT_MEMO_TYPE_KYT_FAIL, [number, number?, number?]];
+
+export class LegacyMintMemoError extends Error {}
+
+/**
+ * Helper that decodes the memo of a ckBTC mint transaction to an object.
+ *
+ * @param memo a Cbor encoded memo.
+ * @throws LegacyMintMemoError when the memo length is 0 or 32 which identifying a legacy memo which are not supported by this help.
+ * @returns {MintMemo} the decoded memo object.
+ */
+export const decodeMintMemo = (memo: Uint8Array | number[]): MintMemo => {
+  // Legacy minting transaction have a memo of length 0 or 32.
+  // We ignore them - those are not supported by this decoder.
+  if (memo.length === 0 || memo.length === 32) {
+    throw new LegacyMintMemoError();
+  }
+
+  return Cbor.decode(new Uint8Array(memo)) as MintMemo;
```

Comment:
This collection of nested arrays is not really suitable as a type to return from a library function.
How about creating a proper struct type with fields for each of the parts?

Reply:
As staten in the description of the PR, this PR just copies the existing code from NNS dapp.

But sure, if we want to go futher, a struct or array, I don't have a strong opinion what's probably more important though is asserting the outcome of the decoding. We can use for example z or do it manually.

Reply:
After thinking about it, I rather like to not create a struct and transmit the object as decided. It aligns better with the general pattern of ic-js (not legacy code).

Regarding the validation of the memo, I have added a validation with zod.

Reply:
> It aligns better with the general pattern of ic-js

What general pattern is that?

Reply:
Not manipulating any types and providing as answer the types defined in candid.
See any index.ts files for reference. 
For example ckbtc

```
export type {
  Account,
  MinterInfo,
  PendingUtxo,
  RetrieveBtcOk,
  RetrieveBtcStatus,
  RetrieveBtcStatusV2,
  Account as WithdrawalAccount,
} from "../candid/minter";
```

Reply:
Unfortunately, this type is not defined in Candid.
It's defined in Rust, [here](https://github.com/dfinity/ic/blob/046de5375825975b57ca3a6f92cd80eaf062f21a/rs/bitcoin/ckbtc/minter/src/memo.rs#L25).
What you're doing here is a bit like returning the Candid serialized bytes instead of the struct-like type.
Since we don't have Candid, I think it's up to us to define the TypeScript type that's the equivalent of the Rust type in the linked file.

However, if all we need to know is whether this is a KYTFail or not, we could keep things simple and just return a boolean (and name the function correspondingly).

Reply:
Yes indeed that way it is consistant with what we expose through Candid. That way one developer that reads the interface can interpret it as if they would read the backend code, this seems like a consistant approach to me.

I do not think there is any particular value at returning a custom struct we would defined.

Therefore I am suggesting the approach I implemented in this PR and do not understand any arguments more valid than another to use a different definition.

Reply:
That said feel free to provide a type defition and z validation if for you it's an absolute must have.

Reply:
> I do not think there is any particular value at returning a custom struct we would defined.

The value is that when you get `[1, 2, 3]` it's unclear what is what and it's easy to index the wrong value, while if you get `{ type: KYTFail, fee: 2}` it's clear what it means.

It wouldn't be a custom struct. It would be the exact equivalent of the Rust struct that was used to serialize the message in the first place.

I'm sorry but returning illegible data is IMO not good enough for a library. If you don't want to change it, you can ask a third person for their opinion.

Reply:
Like I said above, not against you providing you the interface you wish.